File: Golang/Medium/120_Triangle/solution.go
package triangle

type state struct {
	r int
	c int
}

func minimumTotal(triangle [][]int) int {
	return explore(0, 0, triangle, make(map[state]int))
}

func explore(r int, c int, grid [][]int, memo map[state]int) int {
	if !inbound(r, c, grid) {
		return 0
	}

	key := state{r, c}
	if v, exist := memo[key]; exist {
		return v
	}

	left := explore(r+1, c, grid, memo)
	right := explore(r+1, c+1, grid, memo)
	memo[key] = grid[r][c] + min(left, right)
	return memo[key]
}

func inbound(r int, c int, grid [][]int) bool {
	rInbound := r >= 0 && r < len(grid)
	if !rInbound {
		return false
	}
	cInbound := c >= 0 && c < len(grid[r])
	return cInbound
}

File: Python/Medium/120_Triangle/solution.py
class Solution:
    def minimumTotal(self, triangle: list[list[int]]) -> int:
        return Solution.explore(0, 0, triangle, {})


    def explore(r, c, grid, memo):
        if not Solution.inbound(r, c, grid):
            return 0
        
        key = (r, c)
        if key in memo:
            return memo[key]

        left = Solution.explore(r + 1, c, grid, memo)
        right = Solution.explore(r + 1, c + 1, grid, memo)
        memo[key] = grid[r][c] + min(left, right)
        return memo[key]


    def inbound(r, c, grid):
        rInbound = r >= 0 and r < len(grid)
        if not rInbound:
            return False
        cInbound = c >= 0 and c < len(grid[r])
        return cInbound
        